# Resources

[Linear Interpolation](https://en.wikipedia.org/wiki/Linear_interpolation)

[Missing Data Imputation](http://www.stat.columbia.edu/~gelman/arm/missing.pdf)

[7 Steps to Data Preparation](https://www.kdnuggets.com/2017/06/7-steps-mastering-data-preparation-python.html)

https://scikit-learn.org/stable/modules/svm.html

In [1]:
#Import your libraries
import pandas as pd


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather.csv` file. 

#### First import it into a data frame called `austin`.

In [2]:
# Your code here
austin = pd.read_csv("austin_weather.csv")

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [8]:
# Your code here
# austin.dtypes (int for temperature, objectifs for the other ones)
austin.describe()
# 1319 rows 
# high, average and low temperatures in farenheits + range/ quartiles

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [9]:
# Your code here
austin.head()
# actually the "object" columns seem to be numeric as well 

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [10]:
# Your code here
austin["DewPointAvgF"].unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

There is an occurrence of "-" that makes pandas treat it like an object. 

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [11]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [20]:
# Your code here
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [21]:
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [29]:
# Your code here
missing_values= austin.loc[austin.isnull().any(axis=1)]
missing_values

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4     2013-12-25         58        50        41           44.0          40.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
42    2014-02-01         76        66        55           62.0          59.0   
51    2014-02-10         60        48        35           49.0          36.0   
...          ...        ...       ...       ...            ...           ...   
1269  2017-06-12         94        85        75           72.0          68.0   
1288  2017-07-01         98        89        79           76.0          72.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1308  2017-07-21        104        91        77           74.0          68.0   
1313  2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
...            ...                  ...                 ...   
1269          62.0                 87.0                62.0   
1288          66.0                 85.0                62.0   
1304          66.0                 88.0                63.0   
1308          61.0                 85.0                56.0   
1313          61.0                 91.0                58.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
4                   56.0  ...                      30.33   
6                   47.0  ...                      30.39   
7                   36.0  ...                      30.17   
42                  60.0  ...                      29.81   
51                  66.0  ...                      30.15   
...                  ...  ...                        ...   
1269                36.0  ...                      29.89   
1288                38.0  ...                      30.02   
1304                38.0  ...                      29.90   
1308                27.0  ...                      29.95   
1313                25.0  ...                      30.03   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
4                         30.27                 10.0                10.0   
6                         30.34                 10.0                 9.0   
7                         30.04                 10.0                10.0   
42                        29.75                 10.0                10.0   
51                        30.02                 10.0                 8.0   
...                         ...                  ...                 ...   
1269                      29.81                 10.0                10.0   
1288                      29.92                 10.0                 9.0   
1304                      29.84                 10.0                10.0   
1308                      29.86                 10.0                10.0   
1313                      29.93                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
4                    7.0         10.0         2.0         16.0   
6                    7.0          7.0         1.0         11.0   
7                    7.0         10.0         2.0         14.0   
42                   9.0         14.0         6.0         26.0   
51                   4.0         15.0         9.0         23.0   
...                  ...          ...         ...          ...   
1269                 2.0         15.0         7.0         22.0   
1288                 6.0         15.0         7.0         24.0   
1304                10.0         16.0         3.0         23.0   


There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [30]:
# Your code here
austin.shape, missing_values.shape

((1319, 21), (136, 21))

#### Calculate the ratio of missing rows to total rows

In [31]:
# Your code here
missing_values.shape[0]/austin.shape[0]

0.10310841546626232

As you can see, there is a large proportion of missing data (over 9%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [32]:
# Your code here
austin.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [42]:
# Your code here
austin["PrecipitationSumInches"].isna().sum()/austin["PrecipitationSumInches"].shape[0]

0.09401061410159212

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [43]:
# Your code here 
austin.drop(columns="PrecipitationSumInches", inplace=True)

# Print `austin` to confirm the column is indeed removed
austin

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [45]:
# Your code here
austin_fixed = austin.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [50]:
# Your code here
austin_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

In [51]:
austin.isna().sum()

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [52]:
# Your code here:
austin_fixed["Events"].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [53]:
# Your answer:
# 3: fog, rain, and thunderstorm 

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [57]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for i in event_list:
    austin_fixed[i]=0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
0            20.0         4.0         31.0  Rain , Thunderstorm     0    0   
1            16.0         6.0         25.0                          0    0   
2             8.0         3.0         12.0                          0    0   
3            12.0         4.0         20.0                          0    0   
4            10.0         2.0         16.0                          0    0   
...           ...         ...          ...                  ...   ...  ...   
1314         12.0         5.0         21.0                          0    0   
1315         14.0         5.0         20.0                          0    0   
1316         12.0         4.0         17.0                          0    0   
1317         13.0         4.0         20.0                          0    0   
1318         12.0         4.0         20.0                          0    0   

      Rain Thunderstorm  
0        0            0  
1        0            0  
2        0            0  
3        0            0  
4        0            0  
...    ...          ...  
1314     0            0  
1315     0            0  
1316     0            0  
1317     0            0  
1318     0            0  

[1319 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [60]:
# Your code here
for i in event_list:
    austin_fixed[i]= austin_fixed["Events"].str.contains(i)*1

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [65]:
# Your code here
austin_fixed["Rain"].value_counts()

0    942
1    377
Name: Rain, dtype: int64

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [67]:
# Your code here
austin_fixed.drop(columns="Events", inplace=True)
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   57.0  ...                 10.0                 7.0   
1                   43.0  ...                 10.0                10.0   
2                   27.0  ...                 10.0                10.0   
3                   22.0  ...                 10.0                10.0   
4                   56.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                25.0  ...                 10.0                10.0   
1315                20.0  ...                 10.0                10.0   
1316                19.0  ...                 10.0                10.0   
1317                27.0  ...                 10.0                10.0   
1318                22.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 23 columns]

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [97]:
# Your code here
import datetime
from datetime import datetime as dt
from datetime import toordinal

ImportError: cannot import name 'toordinal' from 'datetime' (/Users/tatianadeferaudy/opt/anaconda3/lib/python3.9/datetime.py)

In [101]:
#def convert_date_to_ordinal(date):
#    return dt.strptime(date, '%Y-%M-%d').toordinal()
austin_fixed["Date"] = austin_fixed["Date"].apply(dt.strptime, args=('%Y-%m-%d' ,))

TypeError: strptime() argument 1 must be str, not Timestamp

In [106]:
austin_fixed["Date"] = austin_fixed["Date"].apply(dt.toordinal)

#### Print `austin_fixed` to check your `Date` column.

In [107]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [111]:
# Your code here:
X = austin_fixed[['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]
Y = austin_fixed["TempAvgF"]

In [135]:
Y

0       60
1       48
2       45
3       46
4       50
        ..
1314    89
1315    91
1316    92
1317    93
1318    88
Name: TempAvgF, Length: 1319, dtype: int64

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [109]:
#Your code here:
import sklearn
from sklearn import model_selection
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [140]:
#Your code here:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=.2, random_state=532)
# the random state variable is necessary so that the split is operated at the same place every time 

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [116]:
# Your code here:
austin_fixed.shape[0]*0.8
ts_rows=1056 


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [118]:
# Your code here:
X_ts_train=X.iloc[0:ts_rows, ] 
X_ts_test=X.iloc[ts_rows: ,]

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [119]:
# Your code here:
Y_ts_train=Y.iloc[0:ts_rows, ] 
Y_ts_test=Y.iloc[ts_rows: ,]

# Bonus Challenge 2

As explained in the README, the main purpose of this lab is to show you the typical process of preparing data for machine learning which sometimes takes up 90% of your time. Data cleaning is a valuable skill to learn and you need to be proficient at various techniques including the ones we showed you above as well as others you'll learn in the future. In the real world this skill will help you perform your job successfully and efficiently.

Now that we're done with data praparation, if you want to expeirence what you'll do in the rest 10% of your time, let's make the final leap.

We will use scikit-learn's [*Support Vector Machines*](https://scikit-learn.org/stable/modules/svm.html) to compute the fit of our training data set, the test on our test data set.

#### In the cell below, import `svm` from `sklearn`:

In [122]:
# Your code here
from sklearn import svm
from sklearn.svm import SVC

#### Now, call `svm.SVC.fit()` on `X_train` and `y_train`. Assign the returned value to a variable called `clf` which stands for *classifier*. Then obtain the test score for `X_test` and `y_test` by calling `clf.score()`.

In [141]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.fit_transform(X_test)

In [143]:
# Your code here
clf = svm.SVC() # this assigns the model 

In [144]:
clf.fit(X_train_normalized, Y_train)

SVC()

In [145]:
clf.score(X_test, Y_test)

/Users/tatianadeferaudy/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:432: UserWarning: X has feature names, but SVC was fitted without feature names
  warnings.warn(


0.026515151515151516

#### You now see the model fit score of your test data set. If it's extremely low, it means your selected model is not a good fit and you should try other models.

#### In addition to fitting `X_train`, `y_train`, `X_test`, and `y_test`, you can also fit `X_ts_train`, `y_ts_train`, `X_ts_test`, and `y_ts_test` if you completed Bonus Challenge 1.

In [146]:
# Your code here
clf.fit(X_ts_train, Y_ts_train)

SVC()

In [147]:
clf.score(X_ts_test, Y_ts_test)

0.015209125475285171

#### We hope you have learned a lot of useful stuff in this lab!